src/main/java/com/radixdlt/crypto/ECKeyUtils.java

```diff
@@ -17,95 +17,90 @@
  */
-class ECKeyUtils {
+public class ECKeyUtils {
 
 	private ECKeyUtils() {
 		throw new IllegalStateException("Can't construct");
 	}
 
-	static final String CURVE_NAME = "secp256k1";
-	static final SecureRandom secureRandom = new SecureRandom();
-	private static final X9ECParameters curve;
-	static final ECDomainParameters domain;
-	static final ECParameterSpec spec;
-	private static final byte[] order;
+	private static final String CURVE_NAME = "secp256k1";
+	private static SecureRandom secureRandom;
+	private static X9ECParameters curve;
+	private static ECDomainParameters domain;
+	private static ECParameterSpec spec;
+	private static byte[] order;
 
-	static {
-	    curve = CustomNamedCurves.getByName(CURVE_NAME);
+	public static SecureRandom secureRandom() {
+		System.out.println("ECKeyUtils - static `secureRandom()`");
+		return secureRandom;
+	}
+
+	public static X9ECParameters curve() {
+		return curve;
+	}
 
-	    domain = new ECDomainParameters(curve.getCurve(), curve.getG(), curve.getN(), curve.getH());
-	    spec = new ECParameterSpec(curve.getCurve(), curve.getG(), curve.getN(), curve.getH());
+	public static ECParameterSpec spec() {
+		return spec;
+	}
 
-	    order = adjustArray(domain.getN().toByteArray(), ECKeyPair.BYTES);
+	public static ECDomainParameters domain() {
+		return domain;
+	}
 
-        FixedPointUtil.precompute(curve.getG());
+	static {
+		System.out.println("ECKeyUtils - static `{}`");
+		install();
 	}
 
-	// Must be after secureRandom init
-	static final KeyHandler keyHandler = new BouncyCastleKeyHandler(secureRandom, curve);
+	static synchronized void install() {
+		System.out.println("ECKeyUtils - static `install()`");
+		if (RuntimeUtils.isAndroidRuntime()) {
+			new LinuxSecureRandom();
+			// Ensure the library version of BouncyCastle is used for Android
+			Security.removeProvider(BouncyCastleProvider.PROVIDER_NAME);
+		}
+		Provider requiredBouncyCastleProvider = new BouncyCastleProvider();
+		Provider currentBouncyCastleProvider = Security.getProvider(BouncyCastleProvider.PROVIDER_NAME);
 
-	static byte[] calculateMAC(byte[] salt, byte[] iv, ECPublicKey ephemeral, byte[] encrypted) throws IOException {
-		byte[] ephemeralPubKey = ephemeral.getBytes();
-		int bytesLen = iv.length + encrypted.length + ephemeralPubKey.length;
-		byte[] bytes = new byte[bytesLen];
-		System.arraycopy(iv, 0, bytes, 0, iv.length);
-		System.arraycopy(ephemeralPubKey, 0, bytes, iv.length, ephemeralPubKey.length);
-		System.arraycopy(encrypted, 0, bytes, iv.length + ephemeralPubKey.length, encrypted.length);
-
-		try {
-			Mac mac = Mac.getInstance("HmacSHA256");
-	        mac.init(new SecretKeySpec(salt, "HmacSHA256"));
-	        return mac.doFinal(bytes);
-		} catch (GeneralSecurityException e) {
-			throw new IOException(e);
-	    }
+		// Check if the currently installed version of BouncyCastle is the version
+		// we want. NOTE! That Android has a stripped down version of BouncyCastle
```

Comment:
This is copied over from Java Library, which considered Android. 

Legend:
Let `Y <- X` denote "X is dependent on Y"
JC: Java Common
JL: Java Library
AW: Android Wallet

We have (before the notion of Java Common):
`JL <- AW`

And I'm introducing:
`JC  <- JL <- AW`,

So what ever code we had in `JL` to support Android Wallet, is propagated into `JC`, does that make sense?

Or should that code, in fact, never ever have been part of JL to begin with? I.e. should just be removed?

Reply:
Hmm.  Not an Android expert, but suggest we leave this in place until we can find out whether it is a good idea to remove.  From my sketchy knowledge of crypto primitives on Android, the BouncyCastle provider on Android has been resolved to some degree in recent releases, but it may be that we can support more versions of Android if we retain this.